api_key_stamper/src/lib.rs

```diff
@@ -1,122 +1,383 @@
+    /// If the derived public key doesn't match the passed in value, `PublicKeyMismatch` is returned.
+    pub fn from_bytes<B: AsRef<[u8]>>(
+        private_key_bytes: B,
+        public_key_bytes: Option<B>,
+    ) -> Result<Self, StamperError> {
+        let secret_key = SecretKey::from_bytes(private_key_bytes.as_ref().into())
+            .map_err(|e| StamperError::InvalidPrivateKeyBytes(e.to_string()))?;
+        let signing_key = SigningKey::from(secret_key);
+
+        if let Some(pub_bytes) = public_key_bytes {
+            let expected = VerifyingKey::from_sec1_bytes(pub_bytes.as_ref())
+                .map_err(|e| StamperError::InvalidPublicKeyBytes(e.to_string()))?;
+            let actual = signing_key.verifying_key();
+            if expected != *actual {
+                return Err(StamperError::PublicKeyMismatch(
+                    hex::encode(expected.to_encoded_point(true).as_bytes()),
+                    hex::encode(actual.to_encoded_point(true).as_bytes()),
+                ));
+            }
+        }
+
+        Ok(Self { signing_key })
+    }
 
-    let stamp = ApiStamp {
-        public_key: api_key.public_key_hex.clone(),
-        signature: hex::encode(sig.to_der()),
-        scheme: "SIGNATURE_SCHEME_TK_API_P256".to_string(),
-    };
+    /// Load an API key from hex-encoded strings
+    /// Optionally takes in public a public key to check against, after derivation.
+    /// If the derived public key doesn't match the passed in value, `PublicKeyMismatch` is returned.
+    pub fn from_strings<S: AsRef<str>>(
+        private_key: S,
+        public_key: Option<S>,
+    ) -> Result<Self, StamperError> {
+        let private_key_bytes = hex::decode(private_key.as_ref())
+            .map_err(|e| StamperError::HexDecode(e.to_string()))?;
 
-    let json_stamp = serde_json::to_string(&stamp).unwrap();
+        let public_key_bytes = if public_key.is_some() {
+            Some(
+                hex::decode(public_key.unwrap().as_ref())
+                    .map_err(|e| StamperError::HexDecode(e.to_string()))?,
+            )
+        } else {
+            None
+        };
+
+        Self::from_bytes(private_key_bytes, public_key_bytes)
+    }
+
+    /// Helper to create an API from pre-existing files. This is useful if you used
+    /// the Turnkey CLI (<https://github.com/tkhq/tkcli>) to generate keys.
+    pub fn from_files<P: AsRef<Path>, Q: AsRef<Path>>(
+        private_key_path: P,
+        public_key_path: Q,
+    ) -> Result<Self, StamperError> {
+        let private_key = fs::read_to_string(private_key_path.as_ref()).map_err(|e| {
+            StamperError::Io(private_key_path.as_ref().to_path_buf(), e.to_string())
+        })?;
+        let public_key = fs::read_to_string(public_key_path.as_ref())
+            .map_err(|e| StamperError::Io(public_key_path.as_ref().to_path_buf(), e.to_string()))?;
+
+        let private_key_hex_bytes =
```

Comment:
you could use from_strings here to reuse the code from from_string doing the hex::decode

Reply:
🧠